configfile: "config.json"

rule all:
    input:
        expand("txps/{strain}.transcripts.fa", strain=config["strain"])

rule vcf2vci:
    input: 
        vcf = "snps/{file}.mgp.v5.indels.dbSNP142.normed.vcf.gz",
        fasta = "fasta/Mus_musculus.GRCm38.dna.primary_assembly_bgzip.fa.gz"
    output: "snps/{file}.vci.gz"
    params: 
        vci = "snps/{file}.vci"
    shell: 
        "g2gtools vcf2vci -p 8 -i {input.vcf} -o {params.vci} "
        "-s {wildcards.file} -f {input.fasta}"

rule patch:
    input:
        vci = "snps/{file}.vci.gz",
        fasta = "fasta/Mus_musculus.GRCm38.dna.primary_assembly_bgzip.fa.gz"
    output: "fasta/{file}.patched.fa"
    shell:
        "g2gtools patch -p 8 -i {input.fasta} -c {input.vci} -o {output}"

rule transform:
    input:
        vci = "snps/{file}.vci.gz",
        fasta = "fasta/{file}.patched.fa"
    output: "fasta/{file}.trans.fa"
    shell:
        "g2gtools transform -p 8 -i {input.fasta} -c {input.vci} -o {output}"

rule convert:
    input:
        vci = "snps/{file}.vci.gz",
        gtf = "genes/Mus_musculus.GRCm38.102.gtf",
    output: "genes/{file}.gtf"
    shell:
        "g2gtools convert -i {input.gtf} -c {input.vci} -o {output}"

rule db:
    input: "genes/{file}.gtf"
    output: "genes/{file}.db"
    shell:
        "g2gtools gtf2db -i {input} -o {output}"

rule txps:
    input:
        fasta = "fasta/{file}.trans.fa",
        db = "genes/{file}.db"
    output: "txps/{file}.transcripts.fa"
    shell:
        "g2gtools extract -i {input.fasta} -db {input.db} --transcripts > {output}"
